managers/backup-manager/BackupManager.js

```diff
@@ -0,0 +1,342 @@
+        operation: 'backup',
+        type: backup.type,
+        backup_guid: backup.guid,
+        trigger: backup.trigger,
+        state: 'processing',
+        secret: undefined,
+        agent_ip: undefined,
+        started_at: backupStartedAt,
+        finished_at: null,
+        tenant_id: opts.context ? this.getTenantGuid(opts.context) : args.space_guid
+      })
+      .value();
+
+    let instanceInfo;
+    const result = _
+      .chain(opts)
+      .pick('deployment')
+      .assign({
+        subtype: 'backup',
+        backup_guid: backup.guid,
+        agent_ip: undefined,
+        tenant_id: opts.context ? this.getTenantGuid(opts.context) : args.space_guid,
+        description: `${backup.trigger} backup triggerred by ${data.username} at ${data.started_at}`
+      })
+      .value();
+
+    function createSecret() {
+      return utils
+        .randomBytes(12)
+        .then(buffer => buffer.toString('base64'));
+    }
+
+    function normalizeVm(vm) {
+      let vmParams = _.pick(vm, 'cid', 'agent_id', 'job', 'index');
+      return _.set(vmParams, 'iaas_vm_metadata.vm_id', config.backup.provider.name === CONST.IAAS.AZURE ? vmParams.agent_id : vmParams.cid);
+    }
+
+    let metaUpdated = false,
+      backupStarted = false;
+
+    return Promise
+      .all([
+        createSecret(),
+        this.getDeploymentIps(deploymentName),
+        this.director.getDeploymentVms(deploymentName).map(normalizeVm)
+      ])
+      .spread((secret, ips, vms) => {
+        // set data and backup secret
+        logger.info(`Starting backup on - ${deploymentName}. Agent Ips for deployment - `, ips);
+        data.secret = backup.secret = secret;
+        return this.agent
+          .getHost(ips, 'backup')
+          .tap(agent_ip => {
+            data.agent_ip = result.agent_ip = agent_ip;
+            instanceInfo = _.chain(data)
+              .pick('tenant_id', 'backup_guid', 'instance_guid', 'agent_ip', 'service_id', 'plan_id')
+              .set('deployment', deploymentName)
+              .set('started_at', backupStartedAt)
+              .value();
+            return BackupManager.registerBnRStatusPoller({
+              operation: 'backup',
+              type: backup.type,
+              trigger: backup.trigger
+            }, instanceInfo);
+          })
+          .then(agent_ip => {
+            return this.agent.startBackup(agent_ip, backup, vms);
+          })
+          .then(() => {
+            backupStarted = true;
+            let put_ret = this.backupStore.putFile(data);
+            logger.info(put_ret);
+            const backup_options = _.chain(result)
+              .set('deployment_name', deploymentName)
+              .set('started_at', backupStartedAt)
+              .value();
+            logger.info(`Backup is initiated with the options: `, backup_options);
+            return backup_options;
+          });
+      })
+      .then(backup_options =>
+        eventmesh.server.updateOperationState({
```

Comment:
We should set result first and then set state. It happens that after setting the state apicontroller picked the state and return res with empty response.
Or rather merge these 2 methods `updateOperationState ` `updateOperationResult `

Reply:
Changed the order. will take up consolidating the two functions in the next iteration as it helps in  reducing apiserver calls